iot/iot-device-registry-infinispan/src/main/java/io/enmasse/iot/registry/infinispan/credentials/CacheCredentialService.java

```diff
@@ -0,0 +1,210 @@
+     * Test if a credentials entry is of the requested tenant/device combination.
+     *
+     * @param tenantId The requested tenant id.
+     * @param deviceId The requested device id.
+     * @param credentialsKey The credentials key.
+     * @param json The JSON payload.
+     * @return {@code true} if the entry matches, {@code false} otherwise.
+     */
+    private static boolean isForDevice(String tenantId, String deviceId, CredentialsKey credentialsKey, String json) {
+        if (!credentialsKey.getTenantId().equals(tenantId)) {
+            return false;
+        }
+
+        if (json == null || json.isBlank()) {
+            return false;
+        }
+
+        final JsonObject credentials = new JsonObject(json);
+        var currentDeviceId = credentials.getString(CredentialsConstants.FIELD_PAYLOAD_DEVICE_ID);
+        if (currentDeviceId == null) {
+            return false;
+        }
+
+        return currentDeviceId.equals(deviceId);
+    }
+
+    @Override
+    public void get(String tenantId, String type, String authId, Span span, Handler<AsyncResult<CredentialsResult<JsonObject>>> resultHandler) {
+        get(tenantId, type, authId, null, span, resultHandler);
+    }
+
+    @Override
+    public void get(String tenantId, String type, String authId, JsonObject clientContext, Span span, Handler<AsyncResult<CredentialsResult<JsonObject>>> resultHandler) {
+
+        var key = new CredentialsKey(tenantId, authId, type);
+
+        var future = cache
+                .getAsync(key)
+                .thenApplyAsync(r -> {
+                    if (r != null) {
+                        return CredentialsResult.<JsonObject>from(HTTP_OK, new JsonObject(r));
+                    } else {
+                        return CredentialsResult.<JsonObject>from(HTTP_NOT_FOUND);
+                    }
+                });
+
+        completeHandler(future, resultHandler);
+
+    }
+
+    @Override
+    public void add(String tenantId, JsonObject otherKeys, Handler<AsyncResult<CredentialsResult<JsonObject>>> resultHandler) {
+
+        final String authId = otherKeys.getString(CredentialsConstants.FIELD_AUTH_ID);
+        final String type = otherKeys.getString(CredentialsConstants.FIELD_TYPE);
+
+        var key = new CredentialsKey(tenantId, authId, type);
+
+        add(key, otherKeys, resultHandler);
+
+    }
+
+    private void add(CredentialsKey key, JsonObject payload, Handler<AsyncResult<CredentialsResult<JsonObject>>> resultHandler) {
+
+        var future = cache
+                .withFlags(Flag.FORCE_RETURN_VALUE)
+                .putIfAbsentAsync(key, payload.encode())
+                .thenApplyAsync(r -> {
+                    if (r == null) {
+                        return CredentialsResult.<JsonObject>from(HTTP_CREATED);
+                    } else {
+                        return CredentialsResult.<JsonObject>from(HTTP_CONFLICT);
+                    }
+                });
+
+        completeHandler(future, resultHandler);
```

Comment:
it looks like this may be completed before the async operation creates the result.

Reply:
In that case then the `whenComplete` method will be called directly.